lib/irb/completion.rb

```diff
@@ -128,35 +140,42 @@ def self.retrieve_files_to_require_relative_from_current_dir
       if tok && tok.event == :on_ident && tok.state == Ripper::EXPR_CMDARG
         case tok.tok
         when 'require'
-          result = retrieve_files_to_require_from_load_path.select { |path|
+          result = BaseCompletor.retrieve_files_to_require_from_load_path.select { |path|
```

Comment:
I don't see any benefit of having these `retrieve_*` methods as class methods. Let's define these methods as instance methods too so we can simply call it here and memoize it on this instance.

Reply:
Since completor is created many times, memoizing in instance will make this cache disappear.

Reply:
What's the benefit of 1) initialising a new completer in every new invocation over 2) initialising it once and pass required arguments to it (like #711)?

IMO, using 2) would make memoizing easier to maintain and understand as we can cache the files in its ivars.

Also, using parent class' class methods for caching feels a bit anti-pattern to me 🤔

Reply:
An example of not recreating a completor instance should be like this:
```ruby
# Full spec. Reset when binding is modified.
class Completor
  def reset_cache
    # When binding.eval(something) is executed in IRB,
    # all cache should be invalidated because lvars, ivars and other might be changed.
    # Calculation result might change even if `binding` `preposing` `target` `postposing` is the same.
    # For example, `Readline=Reline` will change the completion canidates of `Readline.a`
  end

  def completion_candidates(preposing, target, postposing, binding)
    cache.fetch([preposing, target, postposing]) { calculate_candidates }
  end

  def doc_namespace(preposing, matched_target, postposing, binding)
    cache.fetch([preposing, target, postposing]) { calculate_doc }
  end
end

# Simplified spec. Reset when one of preposing, target, postposing is changed or binding is modified.
class Completor
  def reset_and_cache(preposing, target, postposing, binding)
    # Always invalidate old cache and cache the parameters
  end

  def completion_candidates
    # Use cached value or recalculate from cached parameters
  end

  def doc_namespace(matched_target)
    # Use cached value or recalculate from cached parameters
  end
end

class Completor
  def completion_candidates(preposing, target, postposing, binding)
    # Should always cache, always invalidate old cache
  end

  def doc_namespace(matched_target)
    # Should use the cache
  end
end

class Completor
  def completion_candidates(preposing, target, postposing, binding)
    # Don't cache
  end

  def doc_namespace(preposing, matched_target, postposing, binding)
    # Always recalculate
  end
end
```

For `reset_cache()` and `reset_and_cache(value_to_cache)` pattern, I think recreating an instance is better than implementing reset for an object.
For caching-in-completion_candidates pattern, there are implicit rules that `completion_candidates` should always invalidate old cache, doc_namespace can always use the cache.

Recreating an instance instead of resetting an instance will simplify it.

Reply:
I don't see any "production" use case for resetting cache. If it's simply for testing, I think we can just memoise with ivar and recreate instances in test teardown. WDYT?

Reply:
Reset is for production. not for test.

```ruby
irb(main):001> a=apple=1
#=> 1
irb(main):002> a            # complete('a','','')        #=> ["a", "apple", "at_exit", ...]
irb(main):002> a.           # complete('a.','','')       #=> ["a.abs", "a.allbits?", ...]
irb(main):002> a.t          # complete('a.t','','')      #=> ["a.to_s", "a.to_int", ...]
irb(main):002> a.to_s       # doc('a.to_s')              #=> "Integer#to_s"
irb(main):002> a=:a;a.t     # complete('a.t','a=:a;','') #=> ["a.to_s", "a.to_sym", "]
irb(main):002> a=:a;a.to_s  # doc('a.to_s')              #=> "Symbol#to_s"
irb(main):002> a=avocado='!'
#=> "!"                     # reset
irb(main):003> a            # complete('a','','')        #=> ["a", "apple", "avocado", ...]
irb(main):003> a.           # complete('a.','','')       #=> ["a.ascii_only?", "a.b", ...]
irb(main):003> a.t          # complete('a.t','','')      #=> ["a.to_s", "a.to_str", ...]
irb(main):003> a.to_s       # doc('a.to_s')              #=> "String#to_i"
```

Reset (or re-initialize) is needed because:
After calculating completion, `doc('a.to_s')` will change from `"Integer#to_s"` to `"Symbol#to_s"`.
After evaluating input, `complete('a','','')` will change from `["a", "apple", "at_exit", ...]` to `["a", "apple", "avocado", ...]`


However, I think we can first implement with not recreating instance, and consider whether to change later. What do you think?

Reply:
Sorry I think I didn't specify what I meant by caching.
I think to match the current behaviour, we only need to cache the result of `retrieve_files_to_require_from_load_path` in ivars?

So I prefer recreating instances with this spec:

```rb
class Completor
  def completion_candidates(preposing, target, postposing, binding)
    # Don't cache
  end

  def doc_namespace(preposing, matched_target, postposing, binding)
    # Always recalculate
  end
end
```

Reply:
:+1:
I'll update this pull request with that later